src/sys/wait.rs

```diff
@@ -215,24 +232,53 @@ fn decode(pid : Pid, status: i32) -> WaitStatus {
     }
 }
 
-pub fn waitpid<P: Into<Option<Pid>>>(pid: P, options: Option<WaitPidFlag>) -> Result<WaitStatus> {
+/// Waits for and returns events that are received from the given supplied process or process group
+/// ID, and associated options.
+///
+/// # Usage Notes
+///
+/// - If the value of PID is greater than `0`, it will wait on the child with that PID.
+/// - If the value of the PID is less than `-1`, it will wait on any child that
+///   belongs to the process group with the absolute value of the supplied PID.
+/// - If the value of the PID is `0`, it will wait on any child process that has the same
+/// group ID as the current process.
+/// - If the value of the PID is `-1`, it will wait on any child process of the current process.
+/// - If the value of the PID is `None`, the value of PID is set to `-1`.
+///
+/// # Possible Error Values
+///
+/// If this function returns an error, the error value will be one of the following:
+///
+/// - **ECHILD**: The process does not exist or is not a child of the current process.
+///   - This may also happen if a child process has the `SIGCHLD` signal masked or set to
+///     `SIG_IGN`.
+/// - **EINTR**: `WNOHANG` was not set and either an unblocked signal or a `SIGCHLD` was caught.
+/// - **EINVAL**: The supplied options were invalid.
+pub fn waitpid<P,O>(pid: P, options: O) -> Result<WaitStatus>
+    where P: Into<Option<Pid>>,
+          O: Into<Option<WaitPidFlag>>
+    {
     use self::WaitStatus::*;
 
     let mut status: i32 = 0;
 
-    let option_bits = match options {
-        Some(bits) => bits.bits(),
-        None => 0
-    };
+    let option_bits = options.into().map_or(0, |bits| bits.bits());
 
-    let res = unsafe { ffi::waitpid(pid.into().unwrap_or(Pid::from_raw(-1)).into(), &mut status as *mut c_int, option_bits) };
+    let res = unsafe {
+        ffi::waitpid(pid.into().unwrap_or (
+            Pid::from_raw(-1)).into(),
+            &mut status as *mut c_int,
+            option_bits
+        )
+    };
 
-    Ok(match try!(Errno::result(res)) {
+    Errno::result(res).map(|res| match res {
         0 => StillAlive,
         res => decode(Pid::from_raw(res), status),
     })
 }
 
+/// Equivalent to `waitpid(-1, None)`, which waits on any child process of the current process.
```

Comment:
It's still confusing because the code doesn't match the comment until you read it a little more thoroughly. I'm a fan of the "principle of least surprise": I was surprised to read that comment and then see that the code didn't match that, though I later realized they're equivalent. I think it'd be better here to just remove the `waitpid...` part and just have a description of what it does. A more descriptive comment below could explain that it's equivalent, but it's definitely not necessary on the title line.